File: q1-25/q25.py
# 2020.07.29
# Problem Statement
# https://leetcode.com/problems/reverse-nodes-in-k-group/
# Looked up the following youtube video:
# https://www.youtube.com/watch?v=cFDvbKTNeCA

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    # do a partial reverse
    def reverse_k(k, first):
        new_tail = first
        prev = None
        for i in range(0 ,k):
            # reconnecting
            next_node = first.next
            first.next = prev
            prev = first
            first = next_node
        new_head = prev
        after = first
        return new_head, new_tail, after
    
    # determine if continue or not
    def has_k_nodes(first, k):
        count = 1
        while first is not None and first.next is not None:
            first = first.next
            count = count + 1
            if count >= k:
                return True
        return False
    
    # function to call
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        # check empty or only 1 node cases
        if head is None: return None
        
        # initialize pointers
        dummy = ListNode(None)
        dummy.next = head
        before = dummy
        first = head
        
        # when still have space to swap
        while Solution.has_k_nodes(first, k):
                
            new_head, new_tail, after = Solution.reverse_k(k, first)

            # reconnecting nodes
            before.next = new_head
            new_tail.next = after
            
            # move forward
            before = first
            first = after
        
        # connect the last (several) node(s)
        before.next = first
        
        return dummy.next

File: tiktok/q25.java
// 2022.10.02
// Problem Statement:
// https://leetcode.com/problems/reverse-nodes-in-k-group/

// idea: seperate to n/k groups and do reverseFirstK for each group,
// everytime move head to reverseFirstK forward by k (old head's next)
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseFirstK(ListNode head, int k) {
        // ex: 1 2 3 4 5, k=4
        // after first recursion get: 1 4 3 2 5
        // want to get: 4 3 2 1 5
        if (k==1) return head;
        ListNode new_tail = head.next; // 2
        ListNode new_head = reverseFirstK(head.next, k-1); // 4
        ListNode temp = new_tail.next;
        new_tail.next = head;
        head.next = temp;
        return new_head;
    }
    public ListNode reverseKGroup(ListNode head, int k) {
        // ex: 1 2 3 4 5 6 7, k=2, curr=1
        // after first function call: 2 1 3 4 5 6 7, new curr=curr->next=3,
        // do reverse starting from 3 to 4, return new head 4, connect with 1 (old current)
        int length = 0;
        ListNode curr = head;
        while (curr!=null) {
            curr = curr.next;
            length++;
        }
        int time = length/k;
        curr = head;
        ListNode prev = curr;
        ListNode ret_head = null;
        for (int i=0; i<time; i++) {
            ListNode temp = reverseFirstK(curr, k);
            if (i!=0) { // connect with the previous part
                prev.next = temp;
            } else { // get the return head
                ret_head = temp;
            }
            // update prev to old curr and move curr forward
            prev = curr;
            curr = curr.next;
        }
        return ret_head;
    }
}